lib/gears/matcher.lua

```diff
@@ -0,0 +1,292 @@
+--- A module to build a set of properties based on a graph of rules.
+--
+-- Rules come from a graph of sources. Each source depend on other and must be
+-- applied before other sources. For this module to work, the graph needs to be
+-- solvable.
+--
+-- @DOC_text_gears_matcher_default_EXAMPLE@
+--
+-- @author Julien Danjou &lt;julien@danjou.info&gt;
+-- @copyright 2009 Julien Danjou
+-- @module gears.matcher
+
+local gtable = require("gears.table")
+local gsort = require("gears.sort")
+local gdebug = require("gears.debug")
+local protected_call = require("gears.protected_call")
+
+local matcher = {}
+
+-- Check if an object matches a rule.
+-- @param o The object.
+-- #tparam table rule The rule to check.
+-- @treturn boolean True if it matches, false otherwise.
+function matcher:_match(o, rule)
+    if not rule then return false end
+    for field, value in pairs(rule) do
+        if o[field] then
+            if type(o[field]) == "string" then
+                if not o[field]:match(value) and o[field] ~= value then
+                    return false
+                end
+            elseif o[field] ~= value then
+                return false
+            end
+        else
+            return false
+        end
+    end
+    return true
+end
+
+-- Check if an object matches any part of a rule.
+-- @param o The object.
+-- #tparam table rule The rule to check.
+-- @treturn boolean True if at least one rule is matched, false otherwise.
+function matcher:_match_any(o, rule)
+    if not rule then return false end
+    for field, values in pairs(rule) do
+        if o[field] then
+            for _, value in ipairs(values) do
+                if o[field] == value then
+                    return true
+                elseif type(o[field]) == "string" and o[field]:match(value) then
+                    return true
+                end
+            end
+        end
+    end
+    return false
+end
+
+--- Does a given rule entry match an object?
+-- @param o The object.
+-- @tparam table entry Rule entry (with keys `rule`, `rule_any`, `except` and/or
+--   `except_any`).
+-- @treturn boolean If `o` matches `entry`.
+function matcher:matches_rule(o, entry)
+    local match = self:_match(o, entry.rule) or self:_match_any(o, entry.rule_any)
+    return match
+        and (not self:_match(o, entry.except))
+        and (not self:_match_any(o, entry.except_any))
+end
+
+--- Get list of matching rules for an object.
+--
+-- If the second `rules` argument is not provided, the rules added with
```

Comment:
```suggestion
-- If the `rules` argument is not provided, the rules added with
```

Reply:
But it only iterates over `rules`?!

Reply:
whoops, forgot that line

It should join all `self._matching_rules`